File: Performance/Graftcode/Client/GraftClient.cs
using GraftServer = graft.nuget.Performance.Graftcode.Server;

namespace Performance.Graftcode.Client;

public static class GraftClient
{
    private const string Host = "ws://localhost/ws";
    private static readonly GraftServer.PerformanceService Service = CreateService();

    public static GraftServer.SmallPayload GetSmall()
    {
        return Service.GetSmall();
    }

    public static GraftServer.LargePayload GetLarge(int sizeMb)
    {
        return Service.GetLarge(sizeMb);
    }

    private static GraftServer.PerformanceService CreateService()
    {
        GraftServer.GraftConfig.Host = Host;
        return new GraftServer.PerformanceService();
    }
}

File: Performance/ReactGraftcode/src/graftClient.ts
import {
  GraftConfig,
  PerformanceService
} from "@graft/nuget-performance.graftcode.server";

const defaultGatewayHost = "ws://localhost:81/ws";

let configuredHost: string | undefined;
let service: PerformanceService | undefined;

export function getGatewayHost() {
  return import.meta.env?.VITE_GRAFTCODE_HOST || defaultGatewayHost;
}

export function configureGraft(host = getGatewayHost()) {
  if (configuredHost === host && service) {
    return service;
  }

  GraftConfig.host = host;
  GraftConfig.rtmCtx = null;
  configuredHost = host;
  service = new PerformanceService();

  return service;
}

export async function getSmallValue() {
  const payload = await callServiceMethod("GetSmall");
  return readField<number>(payload, "Value");
}

export async function getLargePayload(sizeMb: number) {
  const payload = await callServiceMethod("GetLarge", sizeMb);
  const [sizeBytes, body] = await Promise.all([
    readField<number>(payload, "SizeBytes"),
    readField<string>(payload, "Payload")
  ]);

  return {
    sizeBytes,
    payloadLength: body.length
  };
}

async function callServiceMethod(methodName: string, ...args: unknown[]) {
  const serviceInstance = configureGraft().instance;
  return serviceInstance.invokeInstanceMethod(methodName, ...args).execute();
}

async function readField<T>(target: any, fieldName: string): Promise<T> {
  const field = await target.getInstanceField(fieldName).execute();
  return field.getValue() as T;
}
